livekit/src/room/mod.rs

```diff
@@ -1123,34 +1133,48 @@ impl RoomSession {
             }
         });
 
-        let mut participants = self.participants.write();
-        participants.0.insert(sid, participant.clone());
-        participants.1.insert(identity, participant.clone());
+        let mut participants = self.remote_participants.write();
+        participants.insert(identity, participant.clone());
         participant
     }
 
     /// A participant has disconnected
     /// Cleanup the participant and emit an event
     fn handle_participant_disconnect(self: Arc<Self>, remote_participant: RemoteParticipant) {
-        for (sid, _) in remote_participant.tracks() {
+        for (sid, _) in remote_participant.track_publications() {
             remote_participant.unpublish_track(&sid);
         }
 
-        let mut participants = self.participants.write();
-        participants.0.remove(&remote_participant.sid());
-        participants.1.remove(&remote_participant.identity());
+        let mut participants = self.remote_participants.write();
+        participants.remove(&remote_participant.identity());
         self.dispatcher.dispatch(&RoomEvent::ParticipantDisconnected(remote_participant));
     }
 
     fn get_participant_by_sid(&self, sid: &ParticipantSid) -> Option<RemoteParticipant> {
-        self.participants.read().0.get(sid).cloned()
+        self.remote_participants.read().values().find(|x| &x.sid() == sid).cloned()
     }
 
     fn get_participant_by_identity(
         &self,
         identity: &ParticipantIdentity,
     ) -> Option<RemoteParticipant> {
-        self.participants.read().1.get(identity).cloned()
+        self.remote_participants.read().get(identity).cloned()
+    }
+
+    fn set_sid(&self, sid: RoomSid) {
+        let mut guard = self.sid_tx.blocking_lock();
```

Comment:
I don't think we should use a blocking_lock as it could easily lead to deadlock/blocking the tokio runtime.
I feel like using a channel adds too much complexity?
What about implementing a Promise-like struct inside the livekit-protocol. So we can just listen to change to whatever variable. I think this would be cleaner

Reply:
how would we detect the first Room update without a signal of some sort? `handle_room_update` needs to ping the spawned process in some way

Reply:
the Promise utility could stil use a channel internally, but the idea is to hide the implementation as it adds complexity to the Room struct.

Inside the handle_room_update I could see something like:
`self.sid_promise.resolve(my_sid).await`